chcemvediet/apps/anonymization/cron.py

```diff
@@ -25,6 +33,47 @@ def attachment_normalization():
             content_type=content_types.PDF_CONTENT_TYPE
         )
         cron_logger.info(u'Normalized attachment: {}'.format(attachment))
+
+    elif attachment.content_type in content_types.LIBREOFFICE_CONTENT_TYPES:
+        directory_name = tempfile.mkdtemp(dir=u'.')
+        file_copy = os.path.join(directory_name, u'file' + guess_extension(attachment.content_type))
+        shutil.copy2(attachment.file.path, file_copy)
+
+        p = subprocess32.Popen(
+            [u'libreoffice', u'--headless', u'--convert-to', u'pdf', u'--outdir', directory_name,
+             file_copy],
+            stdout=subprocess32.PIPE,
+            stderr=subprocess32.PIPE
+        )
+        try:
+            out, err = p.communicate(timeout=300)
+        except subprocess32.TimeoutExpired:
+            p.terminate()
+            out, err = p.communicate()
+            err += u'Timeout expired.'
+
+        if u'file.pdf' in os.listdir(directory_name):
+            file_pdf = open(os.path.join(directory_name, u'file.pdf'), u'rb')
+            AttachmentNormalization.objects.create(
+                attachment=attachment,
+                successful=True,
+                file=ContentFile(file_pdf.read()),
+                content_type=content_types.PDF_CONTENT_TYPE,
+                debug=u'{}\n{}'.format(out, err)
+            )
+            cron_logger.info(u'Normalized attachment using libreoffice: {}'.format(attachment))
+        else:
+            AttachmentNormalization.objects.create(
+                attachment=attachment,
+                successful=False,
+                content_type=content_types.PDF_CONTENT_TYPE,
+                debug=u'{}\n{}'.format(out, err)
+            )
+            trace = unicode(traceback.format_exc(), u'utf-8')
+            cron_logger.info(u'Normalizing attachment using libreoffice has failed: {}\n{}'
+                             .format(attachment, trace))
+        shutil.rmtree(directory_name)
```

Comment:
Neda sa temp adresar vytvorit context managerom? Aby sme mali garanciu, ze sa zmaze nech by nam cokolvek popadalo v kode. Tj. nieco ako:
```
with funkcia_co_vyraba_temp(...) as directory_name:
    kod_pouzivajuci(directory_name)
```

Reply:
Mozem pouzit takyto model na priecinok. Vo finally klauzule sa zavola shutil.rmtree(). 
```
try:
    directory_name = tempfile.mkdtemp()
except Error_1:
    ...
except Error_2:
    ...
except Exception:
    ...
finally:
    shutil.rmtree(directory_name)
```

Reply:
Funkcne je to spravne, ale imho je to zle citatelne, lebo sa musis nad alokovanym zdrojom (vytvoreny docasny adresar) zamyslat 2x. Raz hore a raz dole. Tym padom je vacsia sanca, ze sa niekde spravi chyba pri upravovani kodu. Ovela robustnejsie je, ked vsetka logika okolo alokovaneho zdroja na jednom mieste. Pouzi radsej [`tempfile.TemporaryDirectory`], co je context manager nad `mkdtemp`. Mas tam aj example ako to pouzit.

[`tempfile.TemporaryDirectory`]: https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory

Reply:
:ok: